File: Breadth-First Search/String/LET_no752_Open the Lock.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        deadends = set(deadends)
        if "0000" in deadends:    # corner case
            return -1
        steps = 0
        queue = collections.deque(["0000"])
        
        while queue:
            for _ in range(len(queue)):
                cur = queue.popleft()
                if cur == target:
                    return steps
                
                for i in range(4):
                    num = int(cur[i])
                    nxt = cur[:i] + str((num + 1) % 10) + cur[i+1:]
                    if nxt not in deadends:
                        deadends.add(nxt)
                        queue.append(nxt)
                    
                    nxt = cur[:i] + str((num - 1) % 10) + cur[i+1:]
                    if nxt not in deadends:
                        deadends.add(nxt)
                        queue.append(nxt)
            steps += 1
        
        return -1

File: Breadth-First Search/String/LET_no752_Open the Lock.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        unordered_set<string> memo(deadends.begin(), deadends.end());
        if (memo.count("0000"))
            return -1;

        int step = 0;
        queue<string> q;
        q.push("0000");

        while (!q.empty()) {
            int n = q.size();
            while (n-- > 0) {
                string cur = q.front();
                q.pop();
                cout << cur << endl;
                if (target.compare(cur) == 0)
                    return step;

                for (int i = 0; i < 4; i++) {
                    char c = cur[i];
                    cur[i] = (c - '0' + 1) % 10 + '0';
                    if (!memo.count(cur)) {
                        memo.insert(cur);
                        q.push(cur);
                    }

                    cur[i] = (c - '0' - 1 + 10) % 10 + '0';
                    if (!memo.count(cur)) {
                        memo.insert(cur);
                        q.push(cur);
                    }
                    cur[i] = c;
                }
            }

            step++;
        }

        return -1;
    }
};